File: app/src/main/java/io/oldering/tvfoot/red/app/domain/matches/BroadcastersAdapter.java
package io.oldering.tvfoot.red.app.domain.matches;

import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import io.oldering.tvfoot.red.R;
import io.oldering.tvfoot.red.app.domain.matches.displayable.BroadcasterRowDisplayable;
import io.oldering.tvfoot.red.databinding.BroadcasterRowBinding;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class BroadcastersAdapter
    extends RecyclerView.Adapter<BroadcastersAdapter.BroadcasterViewHolder> {
  private List<BroadcasterRowDisplayable> broadcasters = new ArrayList<>();

  @Override public BroadcasterViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());

    BroadcasterRowBinding binding =
        DataBindingUtil.inflate(layoutInflater, R.layout.broadcaster_row, parent, false);
    return new BroadcasterViewHolder(binding);
  }

  @Override public void onBindViewHolder(BroadcasterViewHolder holder, int position) {
    holder.bind(broadcasters.get(position));
  }

  @Override public int getItemCount() {
    return broadcasters.size();
  }

  void add(BroadcasterRowDisplayable broadcaster) {
    broadcasters.add(broadcaster);
  }

  public void addAll(Collection<BroadcasterRowDisplayable> broadcasters) {
    this.broadcasters.addAll(broadcasters);
  }

  static class BroadcasterViewHolder extends RecyclerView.ViewHolder {
    private final BroadcasterRowBinding binding;

    BroadcasterViewHolder(BroadcasterRowBinding binding) {
      super(binding.getRoot());
      this.binding = binding;
    }

    public void bind(BroadcasterRowDisplayable broadcaster) {
      binding.setBroadcasterLogoPath(broadcaster.logoPath());
      binding.executePendingBindings();
    }
  }
}

File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/BroadcastersAdapter.kt
package com.benoitquenaudon.tvfoot.red.app.domain.matches

import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.benoitquenaudon.tvfoot.red.R
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.BroadcasterRowDisplayable
import com.benoitquenaudon.tvfoot.red.databinding.BroadcasterRowBinding
import com.benoitquenaudon.tvfoot.red.injection.scope.ActivityScope
import java.util.ArrayList
import javax.inject.Inject

@ActivityScope
class BroadcastersAdapter @Inject constructor(
) : androidx.recyclerview.widget.RecyclerView.Adapter<BroadcastersAdapter.BroadcasterViewHolder>() {
  private val broadcasters = ArrayList<BroadcasterRowDisplayable>()

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BroadcasterViewHolder {
    val layoutInflater = LayoutInflater.from(parent.context)

    val binding = DataBindingUtil.inflate<BroadcasterRowBinding>(layoutInflater,
        R.layout.broadcaster_row, parent, false)
    return BroadcasterViewHolder(binding)
  }

  override fun onBindViewHolder(holder: BroadcasterViewHolder, position: Int) {
    holder.bind(broadcasters[position])
  }

  override fun getItemCount(): Int = broadcasters.size

  // This should not be executed async because we are freezing the layout right afterwards
  fun addAll(broadcasters: Collection<BroadcasterRowDisplayable>) {
    this.broadcasters.addAll(broadcasters)
  }

  class BroadcasterViewHolder(
      private val binding: BroadcasterRowBinding
  ) : androidx.recyclerview.widget.RecyclerView.ViewHolder(binding.root) {

    fun bind(broadcaster: BroadcasterRowDisplayable) {
      binding.broadcasterLogoPath = broadcaster.logoPath
      binding.executePendingBindings()
    }
  }
}
